Latest Post Block: Refactor settings panel to use ToolsPanel

karthick-murugan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Part of: https://github.com/WordPress/gutenberg/issues/67813

Fixes https://github.com/WordPress/gutenberg/issues/67931

## Why?
Latest Post Block code to include ToolsPanel instead of PanelBody.

## How?
Refactored Latest Post Block code to include ToolsPanel instead of PanelBody.

## Testing Instructions

- Login to Dashboard
- Edit any post/ page
- Add Archive Title block.
- Check settings panel of the Latest Post block and verify the change.

## Screenshots or screencast <!-- if applicable -->

**Before**

![before_posts](https://github.com/user-attachments/assets/e7e3a697-7a90-463e-b574-dd60d9a4132b)


**After**

![after_posts](https://github.com/user-attachments/assets/75033f61-8937-4ef9-9aab-47bf4e0bee95)

karthick-murugan committed:
Image size fix in lightbox

karthick-murugan committed:
Revert "Image size fix in lightbox"

This reverts commit 63f81c1e9cbb3b0513abca2b8cb579fef374ed7b.

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Update toolspanel to latest posts block

karthick-murugan committed:
Design changes

fabiankaegy commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,197 +228,279 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+					onDeselect={ () =>
+						setAttributes( { order: 'desc', orderBy: 'date' } )
+					}
+				>
```
Instead of one `ToolsPanelItem` called `Sorting Options` we should have each of the elements in this section as it's own `ToolsPanelItem`. That requires a bit of a larger refactor but we need to ensure each item can be reset on its own.

fabiankaegy reviewed (dismissed):
This one needs a little more refinement. 

1. We should ensure that every single control inside any of the `ToolsPanel` component gets their own `ToolsPanelItem`
2. The UI of the Featured Image settings currently has some spacing issues:
![CleanShot 2024-12-13 at 15 41 48@2x](https://github.com/user-attachments/assets/759f19d4-3ea9-406d-9f15-66d3a279b390)
3. We should ensure that all the defaults that render by default today have the `isShownByDefault` prop added so they cannot get removed.

karthick-murugan committed:
feedback changes

karthick-murugan commented:
@fabiankaegy  - Only fixed the spacing issues on the Featured Image settings. Need some advice on how to proceed for converting ImageSizeControl and QueryControls options to have their own reset and toolspanelitems.

t-hamano reviewed (commented):
Thanks for the PR!

The Latest Posts block controls consist of the following sections:

- Post content
- Post meta
- Featured image (Includes `ImageSizeControl` component)
- Sorting and filtering (includes `QueryControls` component)

The main challenge is that `ImageSizeControl` and `QueryControls` are public components and are not constructed as `ToolsPanelItem`.

As for the `ImageSizeControl` component, we should be able to refactor it with `ResolutionTools` configured with `ToolsPanelItem`.

As for `QueryControls` component, we may need to give up on the individual reset functionality, or rebuild the equivalent functionality on a `ToolsPanelItem` basis and implement it in the Latest Post block.

In any case, it would be too much work to do everything in one PR, so how about just refactoring "Post content" and "Post meta" sections in this PR?

karthick-murugan committed:
Feedback changes

karthick-murugan commented:
Thank you, @t-hamano, for the detailed feedback and suggestions!

I’ve implemented the changes as per your recommendations:

1. Refactored the "Post content" and "Post meta" sections to use ToolsPanelItem for individual reset functionality.
2. Adjusted the logic to ensure seamless functionality and consistency across these sections.

As you suggested, handling ImageSizeControl and QueryControls will require significant effort, so I’ve deferred those changes to keep this PR focused and manageable.

Please let me know if there are any additional adjustments needed or further feedback. I appreciate your guidance throughout this process!

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-							},
-						] }
+				>
+					<ToggleControl
```
I think _each_ control in the Post content panel needs to be wrapped in a `ToolsPanelItem`.

karthick-murugan replied:
@t-hamano - Updated all the changes.

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+				}
+			>
+				<ToolsPanelItem
+					hasValue={ () => displayPostContent !== undefined }
```
```suggestion
					hasValue={ () => !! displayPostContent }
```

karthick-murugan replied:
Changes updated

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+							displayPostContent
+								? 'latest-posts-toggle-control-margin-bottom'
+								: ''
 						}
```
If you wrap each of the three controls in a `ToolsPanelItem`, you won't need this style.

karthick-murugan replied:
Changes updated

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+				}
+			>
+				<ToolsPanelItem
+					hasValue={ () => displayAuthor !== undefined }
```
```suggestion
					hasValue={ () => !! displayAuthor }
```

karthick-murugan replied:
Changes updated

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+						displayPostContentRadio: 'excerpt', // Set default value for the radio button
+						excerptLength: MIN_EXCERPT_LENGTH, // Default length for the excerpt
+					} )
+				}
```
Let's add `dropdownMenuProps`.

karthick-murugan replied:
Changes updated

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+						displayAuthor: false,
+						displayPostDate: false,
+					} )
+				}
```
Let's add `dropdownMenuProps`.

karthick-murugan replied:
Changes updated

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +229,114 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+					/>
+				</ToolsPanelItem>
+				<ToolsPanelItem
+					hasValue={ () => displayPostDate !== undefined }
```
```suggestion
					hasValue={ () => !! displayPostDate }
```

karthick-murugan replied:
Changes updated

karthick-murugan committed:
Feedback updates

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -77,6 +80,8 @@ function getFeaturedImageDetails( post, size ) {
 
 export default function LatestPostsEdit( { attributes, setAttributes } ) {
 	const instanceId = useInstanceId( LatestPostsEdit );
+	const dropdownMenuProps = useToolsPanelDropdownMenuProps(); // Moved inside the component
```
```suggestion
	const dropdownMenuProps = useToolsPanelDropdownMenuProps();
```

I think this comment is redundant.

karthick-murugan replied:
@t-hamano - All changes done. Thank you for the review.

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+				resetAll={ () =>
+					setAttributes( {
+						displayPostContent: false,
+						displayPostContentRadio: 'excerpt', // Set default value for the radio button
+						excerptLength: MIN_EXCERPT_LENGTH, // Default length for the excerpt
```
```suggestion
						displayPostContentRadio: 'excerpt',
						excerptLength: DEFAULT_EXCERPT_LENGTH,
```

We need to change the value of `excerptLength` to the _default_ value. It's a good idea to define anew constant for the default.

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
+						setAttributes( { displayPostContent: false } )
 					}
-				/>
+				>
```
```suggestion
					isShownByDefault
				>
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
 							} )
 						}
-					/>
+					>
```
```suggestion
						isShownByDefault
					>
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-							min={ MIN_EXCERPT_LENGTH }
-							max={ MAX_EXCERPT_LENGTH }
-						/>
+						>
```
```suggestion
							isShownByDefault
						>
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-				/>
-				<ToggleControl
-					__nextHasNoMarginBottom
+				>
```
```suggestion
					isShownByDefault
				>
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-				/>
-			</PanelBody>
-
+				>
```
```suggestion
					isShownByDefault
				>
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
 				{ displayPostContent && (
-					<RadioControl
+					<ToolsPanelItem
+						hasValue={ () => displayPostContentRadio !== undefined }
```
```suggestion
						hasValue={ () => !! displayPostContentRadio }
```

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +232,130 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-								setAttributes( { excerptLength: value } )
+							onDeselect={ () =>
+								setAttributes( {
+									excerptLength: MIN_EXCERPT_LENGTH,
```
```suggestion
									excerptLength: DEFAULT_EXCERPT_LENGTH,
```

karthick-murugan committed:
Feedback changes

t-hamano commented on packages/block-library/src/latest-posts/constants.js:
```diff
@@ -1,3 +1,4 @@
 export const MIN_EXCERPT_LENGTH = 10;
 export const MAX_EXCERPT_LENGTH = 100;
 export const MAX_POSTS_COLUMNS = 6;
+export const DEFAULT_EXCERPT_LENGTH = 10;
```
```suggestion
export const DEFAULT_EXCERPT_LENGTH = 55;
```

This value must be [a default value defined in block.json](https://github.com/WordPress/gutenberg/blob/129587828d6f2908ddf7e1027ecb14f115e095e0/packages/block-library/src/latest-posts/block.json#L34), otherwise it will not reset properly.

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +233,135 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
 				{ displayPostContent && (
-					<RadioControl
+					<ToolsPanelItem
+						hasValue={ () => !! displayPostContentRadio }
```
```suggestion
						hasValue={ () => displayPostContentRadio !== 'excerpt' }
```

Sorry, my previous feedback was incorrect. Here too, we need to check the default value defined in block.json.

t-hamano commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -227,68 +233,135 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
-							__nextHasNoMarginBottom
-							__next40pxDefaultSize
+						<ToolsPanelItem
+							hasValue={ () => excerptLength !== undefined }
```
```suggestion
							hasValue={ () =>
								excerptLength !== DEFAULT_EXCERPT_LENGTH
							}
```

karthick-murugan replied:
@t-hamano - All changes done. Thank You.

karthick-murugan committed:
Feedback changes

t-hamano reviewed (approved):
LGTM! I think everything is working as expected.
